Inline host config

ManasJayanth opened the pull request:
Attempts to close #27

ManasJayanth committed:
Inline npm deps so that GCC can inline

ManasJayanth committed:
Log bundle size

prometheansacrifice commented:
Moving all deps to devDeps since they are all resolved during compile time.

ManasJayanth committed:
Update reconciler to 0.12.0

jquense commented:
Nice! I'm not sure about about the dependency move, what about all the runtime deps like dom helpers and react. Those are still needed at runtime and probably shouldn't be inlined right?

prometheansacrifice commented:
As much as I understand, not all (especially the large ones) functions are
inline

On Sat 16 Jun, 2018, 6:37 PM Jason Quense, <notifications@github.com> wrote:

> Nice! I'm not sure about about the dependency move, what about all the
> runtime deps like dom helpers and react. Those are still needed at runtime
> and probably shouldn't be inlined right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jquense/react-dom-lite/pull/30#issuecomment-397811284>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC9BukH7CRr2Zx2qWY2vGmK_VLRki0_gks5t9QMZgaJpZM4UqVWL>
> .
>

prometheansacrifice commented:
I think the closure compiler takes a call on what to inline or not. And
only small functions, where perf gains from avoiding extra stack frames are
significant, get inlined.

On Sat 16 Jun, 2018, 6:55 PM Jason Quense, <notifications@github.com> wrote:

> I think we only want to inline react-reconciler, otherwise the bundle will
> include React and dom-helpers which will be a big bundle increase for folks
> since they will have duplicated copies of react, and anyone using
> dom-helpers will also be duplicated right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jquense/react-dom-lite/pull/30#issuecomment-397812194>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC9BumBhr5jyhXkCcmRMa6QB35Q1mvwRks5t9QdcgaJpZM4UqVWL>
> .
>

jquense commented:
sure but it will till make the practical bundle size of ppl using RDL bigger if we let it do that. I don't think that's what we want

prometheansacrifice commented:
I think I misunderstood earlier. You're right. We're on the same page. I have re-introduced other deps back in package.json and marked them external for rollup.

ManasJayanth committed:
Inline conditional module loading

ManasJayanth committed:
Use Java version of closure compiler

Inlining was not seen in simple mode

prometheansacrifice commented:
Please compare the build sizes between commit https://github.com/jquense/react-dom-lite/pull/30/commits/4a53decd67804d563658f40f13032856d7870e8f  and <latest one>. This is because previously we didn’t inline npm deps in the cjs bundles. 

I added `fbjs` so that we reuse some of the code thats bundled. It helps dedup some code.

Added index.js so that minified bundle is required too. Previously only unminified bundle was a part of package.json's main.

Had to use ADVANCED mode for closure compiler. We might have to have build level checks.

I had to remove flow defs for the host config too, for the time being. Will revisit tomorrow or so.

gaearon commented:
>Had to use ADVANCED mode for closure compiler

This is pretty hard to get working, I wouldn't recommend. Why do you need it?

prometheansacrifice commented:
$$hostconfig never got inlined inlined in simple mode.

gaearon commented:
It falling down doesn't mean everything works correctly though. I'd be very cautious.

If it doesn't get inlined, maybe we can explore something on *our* side to fix this.

prometheansacrifice commented:
> Hello world app bundled (created using CRA) fell down to 23.76 KB

Sorry. This wasn't meant you justify use of advanced mode. I want to report the changes in build size. Should have updated my previous comment. You're right about the safety of the build.

I did try a couple of things like avoid unnecessary function calls so that gcc can statically analyse hostconfig. Eventually ran out of ideas.

prometheansacrifice commented:
@gaearon Can you please share some ideas to help inline the host config? My initial attempts were to make sure the object is statically analysable. I could have missed though.

prometheansacrifice commented:
I couldn't get the following to inline in simple mode :( Tested it on the [online tool too](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250Afunction%2520log(m)%2520%257B%250A%2520%2520console.log(m)%253B%250A%257D%250Alog(m)%253B%250A%250A)
```js
function log(m) {
  console.log(m);
}
log(m);
```

prometheansacrifice commented:
Just reporting it here. Dan is right about GCC - I got bit by GCC's advanced mode in a project I was working on.

prometheansacrifice commented:
I stand corrected. Figured why GCC didn't inline my code earlier: `log` was declared global. Following does get inlined.

```js
function main(){
  function log(m) {
    console.log(m);
  }

  log('m');
  log('m');
  log('m');
  log('m');
  log('m');
  log('m');
  log('m');

}

main()
```

Inlined code

```js
function main() {
  console.log("m");
  console.log("m");
  console.log("m");
  console.log("m");
  console.log("m");
  console.log("m");
  console.log("m");
}
main();
```

prometheansacrifice commented:
Reverting the advanced level changes shortly

ManasJayanth committed:
Back to using gcc js and simple mode

prometheansacrifice commented:
I observed the following:

a.js

```js
(function main() {

  const config = {
    log(m) {
      console.log(m);
    }
  }

  function reconciler(c) {
    const log = c.log;
    log('jere');
  }

  reconciler(config);

}())

```

a.min.js
```js
(function() {
  (function(a) {
    a = a.log;
    a("jere");
  })({
    log: function(a) {
      console.log(a);
    }
  });
})();
```

b.js
```js
(function main() {

  const config = {
    log(m) {
      console.log(m);
    }
  }

  function reconciler(c) {
    const log = c.log;
    c.log('here');
    // log('jere');
  }

  reconciler(config);

}())
```

b.min.js
```js
(function() {
  console.log("here");
})();
```

ManasJayanth committed:
Experimental: WIP: Adds a babel transform to move the host config

The transform script moves host config inside the reconciler() to help
GCC inline code better

ManasJayanth committed:
Fixes variable renaming bug

prometheansacrifice commented:
# Explanation

## Trying to understand GCC's behaviour

I tried replicating the bundle's code behaviour of wrapping a config object and tried running GCC on it. 

I got GCC to inline the following:

```js
  const config = {
    log(m) {
      console.log(m);
    }
  }

  function reconciler(c) {
    const log = c.log;
    c.log('here');
    // log('jere');
  }

  reconciler(config);
```
But not the following which resembles the bundle better

```js
  const config = {
    log(m) {
      console.log(m);
    }
  }

  function reconciler(c) {
    const log = c.log;
    log('jere');
    return {
      foo() {
        // More side effects
      }
    }
  }

  reconciler(config);
```

I tried a couple of other variations too. Reported it on [SO](https://stackoverflow.com/questions/51383149/closure-compiler-how-does-it-decided-when-to-inline) and [Google Groups](https://groups.google.com/forum/#!topic/closure-compiler-discuss/x-i9WIqV4cM). [Tweeted](https://twitter.com/ManasJayanth/status/1019220997998764034) about it too.

Most reliable [response](https://stackoverflow.com/a/51465397/2065736) was from one of the maintainers of GCC on SO.

> There is no one answer. The compiler uses heuristics and side effect calculations to decide when to inline. The compiler is also less likely to inline in the global scope than in nested scopes.

Thats when I decided to transform to the structure of our bundle to resemble React DOM's.

## The babel transform

Basically, the scripts transforms,

```js
  function createInstance() {
  }
  function appendChild() {
  }
  const config = {
    createInstance: createInstance
    appendChild: appendChild
  }
  function reconciler(config) {
    var ax = config.createInstance,
          by = config.appendChild;

   // ...          
  }
  reconciler(config);
```
to 

```js
  // const config = { ... }  is removed
  function reconciler(config) {
      var ax = function createInstance() {
      }
      var by = function appendChild() {
      }
      // ...          
  }
  reconciler({});
```


## Gotchas
The transform depends on the line `Reconciler(__DEV__ ? HostConfigDev : HostConfigProd)` in the code. This is because of the `reconciler(config);` to `reconciler({});` transform.

I figure what `Reconciler(__DEV__ ? HostConfigDev : HostConfigProd)` looks like in the bundle using mozilla's sourcemap consumer lib, and remove it from the ast and perform the rest of the operations.

I am still inspecting both transformed and the inlined code (using the newly added debug mode in GCC) and see if I can spot any potential issues. Another set of eyes would really help.

## Will this transform work unconditionally?
Two things are absolutely necessary,

1. Any change to `Reconciler(__DEV__ ? HostConfigDev : HostConfigProd)` must be reflected in the transform script.
2. Config must to written in modules so that Rollup declares them as functions and const/let bindings. The transformer reliable on VariableDeclarations and FunctionDeclarations. Host configs must not longer be of the form { shouldSetTextContent(){ ... }, createInstance() {...} } as the transformer doesn't account for them
 


Please let me know your thoughts.

jquense commented:
@prometheansacrifice what is the practical benefit we are getting here again, do we know how many bytes were saving this way? It's kinda convoluted so i'm not sure the added fragility to the build is worth it unless its a significant savings yeah?

jquense commented on flow-typed/react-reconciler.js:
```diff
@@ -60,175 +60,175 @@ declare module 'react-reconciler' {
-      text: string,
-    ): void,
-  };
+  // declare export type HostConfig<T, P, I, TI, HI, PI, C, CC, CX, PL> = {
```
why are they commented out?

prometheansacrifice replied:
Those types, particularly ssr and mutation blocks need to be flattened.

jquense commented on package.json:
```diff
@@ -69,10 +78,9 @@
-    "react": "^16.3.0",
-    "react-reconciler": "0.10.0",
-    "warning": "^3.0.0"
+    "fbjs": "^0.8.17"
```
i really don't want a dep on fbjs if we can help it, its more of an internal FB project, and REact itself is moving away from including it

prometheansacrifice replied:
Got it. It helped us avoid duplicate declarations for warning and other modules.

jquense replied:
yeah i wonder if it's worth making out own one for that? I have a package react-tackle-box i dump random common stuff into, if it's helpful i could put stuff there

prometheansacrifice commented:
@jquense Did you get a chance to look at the GCC'd bundle?
